### PR TITLE
Feature/unr 2762 latency tracking improvements

### PR DESCRIPTION
**Pull request by improbable-danny**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Tracing interface improvements so the interface/usage is not so cluttered. Also added entity creation tracing via the EntityFactory

#### Tests
Tested using the test gym

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 1]
```diff
@@ -50,13 +62,13 @@ bool ComponentFactory::FillSchemaObject(Schema_Object* ComponentObject, UObject*
...
+			if (LatencyTracer != nullptr && OutLatencyTraceId != nullptr)
 			{
-				OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Cmd.Property);
-				if (OutLatencyTraceId == USpatialLatencyTracer::InvalidTraceKey)
+				*OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Cmd.Property);
+				if (*OutLatencyTraceId == USpatialLatencyTracer::InvalidTraceKey)
 				{
 					// Check for sending a nested property
-					OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
+					*OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: As I understand, `OutLatencyTraceId` will be equal to the last value we process in that loop? All the previous values are discarded?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 2]
```diff
@@ -130,9 +142,9 @@ bool ComponentFactory::FillHandoverSchemaObject(Schema_Object* ComponentObject,
 		const uint8* Data = (uint8*)Object + PropertyInfo.Offset;
 
 #if TRACE_LIB_ACTIVE
-		if (LatencyTracer != nullptr)
+		if (LatencyTracer != nullptr && OutLatencyTraceId != nullptr)
 		{
-			OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, PropertyInfo.Property);
+			*OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, PropertyInfo.Property);
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: It's the same case as in `FillSchemaObject`.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp (hunk 1 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Yeah that's true, I will add some code to end traces which are about to be overwritten - we can only send 1 trace with the Schema object at a time. Perhaps I could extend it to send more!

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: If this issue is not resolved in this PR - could you create a jira ticket and add a `// TODO: UNR-<ticket number>` with a short description here and in the case below?

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: We're going to terminate the traces except one in this situation, as it's a bit tricky to manage sending multiple traces!

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: If that's going to be the intentional behaviour - could you put a comment that says so?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 3]
```diff
@@ -598,11 +598,25 @@ void USpatialWorkerConnection::ProcessOutgoingMessages()
 		{
 			FCreateEntityRequest* Message = static_cast<FCreateEntityRequest*>(OutgoingMessage.Get());
 
+#if TRACE_LIB_ACTIVE
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Nit: it seems we can have fewer lines of code duplicated. i.e. have just one call to `Worker_Connection_SendCreateEntityRequest`.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/EntityFactory.h [hunk 4]
```diff
@@ -4,6 +4,7 @@
  
 #include "Core.h"
  
+#include <SpatialCommonTypes.h>
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/EntityFactory.h]:
```suggestion
#include "SpatialCommonTypes.h"
```

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/EntityFactory.h]: We usually include with <> if it's a Worker header, and use "" otherwise.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp [hunk 5]
```diff
@@ -334,12 +334,20 @@ void USpatialLatencyTracer::OnEnqueueMessage(const SpatialGDK::FOutgoingMessage*
...
 		const SpatialGDK::FAddComponent* ComponentAdd = static_cast<const SpatialGDK::FAddComponent*>(Message);
-		WriteToLatencyTrace(ComponentAdd->Trace, TEXT("Moved componentAdd to Worker queue"));
+		WriteToLatencyTrace(ComponentAdd->Data.Trace, TEXT("Moved componentAdd to Worker queue"));
+	}
+	else if (Message->Type == SpatialGDK::EOutgoingMessageType::CreateEntityRequest)
+	{
+		const SpatialGDK::FCreateEntityRequest* CreateEntityRequest = static_cast<const SpatialGDK::FCreateEntityRequest*>(Message);
+		for (auto& Component : CreateEntityRequest->Components)
+		{
+			EndLatencyTrace(Component.Trace, TEXT("Moved createEntityRequest to Worker queue"));
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: Is it intentional, that there's `EndLatencyTrace` and not `WriteToLatencyTrace`?

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLatencyTracer.cpp]: No, good spot!

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp (hunk 3 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: Updated this

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 6]
```diff
@@ -598,9 +598,23 @@ void USpatialWorkerConnection::ProcessOutgoingMessages()
...
+			// We have to unpack these as Worker_ComponentData is not the same as FWorkerComponentData
+			TArray<Worker_ComponentData> UnpackedComponentData;
+			UnpackedComponentData.SetNum(Message->Components.Num());
+			for (int i = 0, Num = Message->Components.Num(); i < Num; i++)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: These are both arrays, can we just append? Or better yet, move the data?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp [hunk 7]
```diff
@@ -598,9 +598,23 @@ void USpatialWorkerConnection::ProcessOutgoingMessages()
...
+			uint32 ComponentCount = UnpackedComponentData.Num();
+#else
+			Worker_ComponentData* ComponentPtr = Message->Components.GetData();
+			uint32 ComponentCount = Message->Components.Num();
+#endif
 			Worker_Connection_SendCreateEntityRequest(WorkerConnection,
-				Message->Components.Num(),
-				Message->Components.GetData(),
+				ComponentCount,
+				ComponentPtr,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: ComponentData?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 8]
```diff
@@ -123,24 +123,12 @@ void USpatialSender::SendAddComponent(USpatialActorChannel* Channel, UObject* Su
...
-		Worker_ComponentData& ComponentData = SubobjectDatas[i];
-		TraceKey LatencyKey = USpatialLatencyTracer::InvalidTraceKey;
-
-#if TRACE_LIB_ACTIVE
-		LatencyKey = TraceKeys[i];
-#endif
-
-		Connection->SendAddComponent(Channel->GetEntityId(), &ComponentData, LatencyKey);
+		FWorkerComponentData& ComponentData = SubobjectDatas[i];
+		Connection->SendAddComponent(Channel->GetEntityId(), &ComponentData);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: So elegant!

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 9]
```diff
@@ -50,14 +62,21 @@ bool ComponentFactory::FillSchemaObject(Schema_Object* ComponentObject, UObject*
...
+				PropertyKey = LatencyTracer->RetrievePendingTrace(Object, Cmd.Property);
+				if (PropertyKey == USpatialLatencyTracer::InvalidTraceKey)
 				{
 					// Check for sending a nested property
-					OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
+					PropertyKey = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
 				}
+				// If we have already got a trace for this actor/component, we will end one of them here
+				if (*OutLatencyTraceId != USpatialLatencyTracer::InvalidTraceKey)
+				{
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Perhaps worth adding a warning here also

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 10]
```diff
@@ -50,14 +62,21 @@ bool ComponentFactory::FillSchemaObject(Schema_Object* ComponentObject, UObject*
...
+					LatencyTracer->EndLatencyTrace(*OutLatencyTraceId, TEXT("Multiple actor component traces not supported"));
+				}
+				*OutLatencyTraceId = PropertyKey;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Wait, won't this override any trace that was active with an invalid trace key?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 11]
```diff
@@ -130,9 +149,14 @@ bool ComponentFactory::FillHandoverSchemaObject(Schema_Object* ComponentObject,
...
 
 #if TRACE_LIB_ACTIVE
-		if (LatencyTracer != nullptr)
+		if (LatencyTracer != nullptr && OutLatencyTraceId != nullptr)
 		{
-			OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, PropertyInfo.Property);
+			// If we have already got a trace for this actor/component, we will end one of them here
+			if (*OutLatencyTraceId != USpatialLatencyTracer::InvalidTraceKey)
+			{
+				LatencyTracer->EndLatencyTrace(*OutLatencyTraceId, TEXT("Multiple actor component traces not supported"));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Likewise warning here

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 12]
```diff
@@ -294,110 +318,84 @@ void ComponentFactory::AddProperty(Schema_Object* Object, Schema_FieldId FieldId
...
-	Worker_ComponentData ComponentData = {};
+	FWorkerComponentData ComponentData = {};
 	ComponentData.component_id = ComponentId;
 	ComponentData.schema_type = Schema_CreateComponentData();
 	Schema_Object* ComponentObject = Schema_GetComponentDataFields(ComponentData.schema_type);
 
 	// We're currently ignoring ClearedId fields, which is problematic if the initial replicated state
 	// is different to what the default state is (the client will have the incorrect data). UNR:959
-	FillSchemaObject(ComponentObject, Object, Changes, PropertyGroup, true, OutLatencyTraceId);
+	FillSchemaObject(ComponentObject, Object, Changes, PropertyGroup, true, GetTraceKeyFromWorkerType(ComponentData));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: `GetTraceKeyFromWorkerType` name confuses me. `GetTraceKeyFromComponentData` ?

File: SpatialGDK/Source/SpatialGDK/Public/SpatialCommonTypes.h [hunk 13]
```diff
@@ -16,6 +16,7 @@ using VirtualWorkerId = uint32;
 using PhysicalWorkerName = FString;
 using ActorLockToken = int64;
 using TraceKey = int32;
+constexpr TraceKey InvalidTraceKey{ -1 };
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialCommonTypes.h]: This is fine, but can you remove the one in SpatialLatencyTracer ?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp (hunk 6 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialWorkerConnection.cpp]: They are POD types so even moving would result in copy :(

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp (hunk 10 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: +1

File: SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp [hunk 14]
```diff
@@ -50,13 +62,24 @@ bool ComponentFactory::FillSchemaObject(Schema_Object* ComponentObject, UObject*
...
 					// Check for sending a nested property
-					OutLatencyTraceId = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
+					PropertyKey = LatencyTracer->RetrievePendingTrace(Object, Parent.Property);
+				}
+				if (PropertyKey != InvalidTraceKey)
+				{
+					// If we have already got a trace for this actor/component, we will end one of them here
+					if (*OutLatencyTraceId != InvalidTraceKey)
+					{
+						UE_LOG(LogSpatialNetSerialize, Warning, TEXT("%s property trace being dropped because too many active on this actor (%s)"), *Cmd.Property->GetName(), *Object->GetName());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentFactory.cpp]: Use `LogComponentFactory`
Driveby, can you fix other misuses of this log type in this file?
